File: examples/javascript/src/webhooks.js
/**
 * Wekeza API Webhooks Module
 * Handles webhook signature verification and event processing
 */

const crypto = require('crypto');

class WekezaWebhooks {
  constructor(webhookSecret) {
    this.webhookSecret = webhookSecret;
  }

  /**
   * Verify webhook signature
   * @param {string} payload - Raw request body
   * @param {string} signature - X-Wekeza-Signature header value
   * @returns {boolean} True if signature is valid
   */
  verifySignature(payload, signature) {
    if (!signature) {
      return false;
    }

    const expectedSignature = crypto
      .createHmac('sha256', this.webhookSecret)
      .update(payload)
      .digest('hex');

    // Use timing-safe comparison to prevent timing attacks
    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expectedSignature)
    );
  }

  /**
   * Parse webhook event
   * @param {string} payload - Raw request body
   * @param {string} signature - X-Wekeza-Signature header value
   * @returns {object} Parsed event data
   * @throws {Error} If signature is invalid
   */
  parseEvent(payload, signature) {
    if (!this.verifySignature(payload, signature)) {
      throw new Error('Invalid webhook signature');
    }

    try {
      return JSON.parse(payload);
    } catch (error) {
      throw new Error('Invalid webhook payload: not valid JSON');
    }
  }

  /**
   * Handle webhook event based on type
   * @param {object} event - Parsed event data
   * @param {object} handlers - Event handlers by type
   * @returns {Promise<any>} Handler result
   */
  async handleEvent(event, handlers) {
    const eventType = event.type || event.event_type;
    
    if (!eventType) {
      throw new Error('Event type not specified in webhook payload');
    }

    const handler = handlers[eventType];
    
    if (!handler) {
      console.warn(`No handler registered for event type: ${eventType}`);
      return null;
    }

    try {
      return await handler(event.data || event.payload);
    } catch (error) {
      console.error(`Error handling ${eventType} event:`, error);
      throw error;
    }
  }
}

module.exports = WekezaWebhooks;

File: examples/python/wekeza_sdk/webhooks.py
"""
Wekeza API Webhooks Module
Handles webhook signature verification and event processing
"""

import hmac
import hashlib
import json
from typing import Dict, Any, Callable


class WebhookVerificationError(Exception):
    """Raised when webhook signature verification fails"""
    pass


class InvalidWebhookPayloadError(Exception):
    """Raised when webhook payload is invalid"""
    pass


class WekezaWebhooks:
    """Handles webhook signature verification and event processing"""
    
    def __init__(self, webhook_secret: str):
        self.webhook_secret = webhook_secret
    
    def verify_signature(self, payload: str, signature: str) -> bool:
        """
        Verify webhook signature
        
        Args:
            payload: Raw request body
            signature: X-Wekeza-Signature header value
            
        Returns:
            bool: True if signature is valid
        """
        if not signature:
            return False
        
        expected_signature = hmac.new(
            self.webhook_secret.encode('utf-8'),
            payload.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()
        
        # Use timing-safe comparison to prevent timing attacks
        return hmac.compare_digest(signature, expected_signature)
    
    def parse_event(self, payload: str, signature: str) -> Dict[str, Any]:
        """
        Parse webhook event
        
        Args:
            payload: Raw request body
            signature: X-Wekeza-Signature header value
            
        Returns:
            Dict containing parsed event data
            
        Raises:
            WebhookVerificationError: If signature is invalid
            InvalidWebhookPayloadError: If payload is not valid JSON
        """
        if not self.verify_signature(payload, signature):
            raise WebhookVerificationError("Invalid webhook signature")
        
        try:
            return json.loads(payload)
        except json.JSONDecodeError as e:
            raise InvalidWebhookPayloadError(f"Invalid webhook payload: not valid JSON - {str(e)}")
    
    def handle_event(self, event: Dict[str, Any], handlers: Dict[str, Callable]) -> Any:
        """
        Handle webhook event based on type
        
        Args:
            event: Parsed event data
            handlers: Event handlers by type
            
        Returns:
            Handler result
            
        Raises:
            InvalidWebhookPayloadError: If event type is missing
        """
        event_type = event.get('type') or event.get('event_type')
        
        if not event_type:
            raise InvalidWebhookPayloadError("Event type not specified in webhook payload")
        
        handler = handlers.get(event_type)
        
        if not handler:
            print(f"Warning: No handler registered for event type: {event_type}")
            return None
        
        try:
            event_data = event.get('data') or event.get('payload')
            return handler(event_data)
        except Exception as e:
            print(f"Error handling {event_type} event: {str(e)}")
            raise
